README
This is PARI/GP, version 2.7.x (*).

PARI/GP Number Theory-oriented Computer Algebra System
        Copyright (C) 2000-2014 The PARI Group, Bordeaux.

==========================================================================

To get started, run 

  tex doc/INSTALL.tex 

and have a look at the file doc/INЅTALL.dvi, extracted verbatim from Appendix A
of the User Manual. If you are in a hurry, look at INSTALL (ascii file). But
a lot of information is missing there.

Once 'Configure' has been run, 'make doc' typesets separate user's manual for
GP and libpari, a tutorial and reference card in directory doc.

See the files NEW and COMPAT for new features and incompatible changes
respectively between 2.7 and older versions. See CHANGES-* for a terse
description of subsequent patches.

For the adventurous: to check out bleeding-edge development versions of
PARI/GP, see the file README-git.

Afterwards, you're on your own.

==========================================================================

There are three mailing lists devoted to the PARI/GP package, and most
feedback should be directed to those. They are:

* pari-announce: to announce major version changes. You can't write to this
one, but you should probably subscribe.

* pari-dev: for everything related to the development of PARI, including
suggestions, technical questions, bug reports or patch submissions.

* pari-users: for everything else.

To subscribe, send empty messages respectively to

   pari-announce-request@pari.math.u-bordeaux.fr
   pari-users-request@pari.math.u-bordeaux.fr
   pari-dev-request@pari.math.u-bordeaux.fr

with a Subject: containing the word "subscribe".

If you are not a member of any of those lists and do not want to become
one, you can write to us at

  pari@math.u-bordeaux.fr

We may forward your mail to the lists above and will definitely try to
correct faulty behaviour, if necessary. But we cannot promise that you
will get an individual answer.

Last but not least, PARI home page can be found at

  http://pari.math.u-bordeaux.fr/

  Thanks for your support, and have fun !

========================================================================

PARI/GP is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version. It is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY WHATSOEVER.

Check the License for details. You should have received a copy of it, along
with the package; see the file 'COPYING'. If not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


(*) Version numbers have the shape: MAJOR.MINOR.patchlevel.status. The status
increases through {alpha, beta, release}. New features are introduced during
the `alpha' phase, tested during `beta', and declared stable when `release'
is reached.
The MAJOR version number will not change for a while.
The MINOR version number increases by 2 when a new status cycle starts (as
soon as a 'release' version is out): odd numbers are reserved for 'released'
versions, even ones for alpha/beta phase.
The patchlevel goes up each time we feel an upgrade is necessary (improvement
or bugfix). 'Release' versions may be updated with further patchlevels if
important bugs need to be fixed before the next cycle reaches completion.


CHANGES
# $Id$
Bug numbers refer to the BTS at http://pari.math.u-bordeaux.fr/Bugs/

Done for version 2.7.4 (released ??/??/2015):
[last column crossreferences current development release 2.8.0]

  Fixed
    1- divisors([2,1]) -> SEGV                                   [#1664] [F84]
    2- acos([Pol(1)]) -> GC bug [#1663]                                  [F85]
    3- primes([1,Pol(2)]) -> SEGV [#1668]                                [F87]
    4- buffer overflow after default(format,"f.precision")               [F95]
    5- qfminim(A, 0, ...) -> stack overflow                      [#1682] [F97]
    6- div_scal_rfrac could create an invalid t_POL              [#1651] [F99]
    7- polroots(t_POL with leading coeff = 0) -> fp exception    [#1690] [F100]
    8- \r cannot deal with very long filenames                   [#1616] [F101]
    9- rnfisabelian(nf, non monic t_POL) -> SEGV                 [#1693] [F102]
   10- possibly incorrect result in nfdisc(T,listP)                      [F104]
   11- wrong dim(Ker) returned by ZM_pivot => SEGV in Z-linear algebra
       routines.  E.g. setrand(1);quadclassunit(-612556842419)   [#1700] [F105]
   12- moebius(factor(18)) -> 1 instead of 0                     [#1702] [F106]
   13- ispower(-167^10) => domain error                          [#1703] [F107]
   14- ispowerful(factor(0)) != ispowerful(0)                            [F108]
   15- expm1(2*I) => wrong result                                        [F109]
   16- gamma(1+a*x+O(x^2)) => error [#1707]                              [F110]
   17- printsep() printed its argument in random format instead of f_RAW [F111]
   18- nfdisc(x^10 - 29080*x^5 - 25772600) -> oo loop            [#1710] [F112]

  Added
ED  1- mingw64 support                                                   [A111]

Done for version 2.7.3 (released 18/02/2015):
[last column crossreferences current development release 2.8.0]

  Fixed
    1- [from 2.7.2] ??bnrL1 and ??ellL1 did not work anymore
    2- thue(f^e*g, ...), e even, (f,g)=1 missed solutions such that f<0   [F22]
XR  3- bnrL1(bnrinit(bnfinit(x^2-168),[6,[1,1]],1)) -> bug in ArtinNumber [F55]
    4- nfisincl(t_POL, t_POL) could lead to wrong negative results        [F58]
    5- polresultant(1+x*z^2,1+y*z^4,z) -> GC error [#1614]                [F59]
BA  6- ellcard over non-prime fields could return wrong results           [F60]
    7- [libpari] FpX_roots could produce GC errors [#1618]                [F61]
    8- weber(1+I) was missing its imaginary part                          [F62]
    9- (1+I)*(1+1/2*I) => wrong result (type errors) [#1619]              [F63]
   10- contfracpnqn([a]) => [1,a;0,1] instead of [a,1;1,0]                [F64]
   11- primes([2^50, 2^50+200000]) => stack overflow                      [F65]
   12- issquare((x+1/2)^2,&z); z => 1.0*x+0.5 instead of x+1/2            [F66]
   13- possibly missing roots in nfroots (when using Trager)              [F68]
   14- quadray(bnf, ideal) did not work                                   [F69]
   15- thue(-14*x^3 + 10*x^2 + 63*x - 5,1) -> "short continued fraction"  [F70]
   16- thue(29*x^3+130*x^2-35*x-48,1) -> "round error" bug                [F71]
   17- T=thueinit(10*x^3+6*x^2-41*x+8,1); thue(T,8) => SEGV [#1630]       [F72]
   18- ellrootno(e,p = 2 or 3) when e not minimal at p => random result   [F73]
   19- agm(1,2+O(5)) -> SEGV [#1645]                                      [F76]
BA 20- [cygwin64] ellap(ellinit([0,0,1,-1,0]),10007) was broken           [F77]
   21- primes([-5,5]) -> [5] (spurious absolute values)                   [F78]
   22- matqr([;]) -> crash                                                [F79]
   23- Fp_rem_mBarrett could return a non-normalized result               [F80]
       p=436^56-35;Mod(271,p)^((p-1)/2) -> p+1
   24- plotcopy would corrupt "string" objects (ROt_ST)                   [F81]
BA 25- [GP] default arguments to GP functions could cause corruption      [F82]
VBr26- [darwin] remove obsolete linker options that cause crashes [#1623] [F83]

Done for version 2.7.2 (released 19/09/2014):
[last column crossreferences current development release 2.8.0]

  Fixed
    1- gaffsg(0, t_PADIC): wrong valuation                                [F21]
    2- (t_INTMOD with word-sized modulus)^(huge negative power) [#1584]   [F24]
    3- (gp -p N) or (primelimit=N in gprc_ for N >= 436273290 resulted in an
       incorrect primetable. N.B. Such commands are now useless: needed primes
       are produced dynamically anyway.                                   [F25]
    4- monomial(exact zero, d, v) returned an invalid t_POL / t_RFRAC     [F26]
    5- contfracpnqn(v, n) returned partial quotients p[-1]/q[-1] ...
       p[n-1]/q[n-1], instead of the documented p[0]/q[0] ... p[n]/q[n]   [F27]
    6- factor((3+4*I)/25) -> factor 2+I had 0 exponent [#1586]            [F29]
BA  7- iferr() could crash if some component of the t_ERROR were clones.  [F31]
    8- nffactor() could overflow the stack when default accuracy too low  [F32]
BA  9- obsolete use of E=[a1,a2,a3,a4,a6] in ellmul crashed  [#1589]      [F33]
   10- incorrect rounding in mulrr/divrr for one-word precision reals     [F34]
BA 11- multiif did not handle correctly return() in conditions [#1590]    [F35]
   12- [0..5] -> [0,0,0,0,0] on some architectures                        [F36]
   13- is_gener_Fp could return wrong results                             [F37]
   14- Fq_sqrtn(t_INT,..,&zeta) could return a wrong root of 1            [F38]
   15- bnfinit: SEGV due to precision issues [#1592]                      [F39]
   16- zm_zc_mul only worked for square zm matrices                       [F40]
   17- genus2red(0,27*x^5+97*x^4+118*x^3+60*x^2+13*x+1,3) -> bug [#1596]  [F41]
   18- [gphelp] oo loop when $COLUMNS too small [#1594]                   [F42]
   19- genus2red(x,-x^6-3*x^4-10*x^2-1,3) -> impossible inverse [#1597]   [F43]
   20- factoru(1) returned a t_MAT instead of the expected "matsmall"     [F44]

src/gp/gp_init.c
/* $Id$

Copyright (C) 2000  The PARI group.

This file is part of the PARI/GP package.

PARI/GP is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation. It is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY WHATSOEVER.

Check the License for details. You should have received a copy of it, along
with the package; see the file 'COPYING'. If not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA. */

/*******************************************************************/
/*                                                                 */
/*                    GP-SPECIFIC FUNCTIONS                        */
/*                                                                 */
/*******************************************************************/
#include "pari.h"
#include "paripriv.h"
#include "../graph/rect.h"
#include "gp.h"

static void whatnow0(char *s) { whatnow(pariOut, s,0); }

static void
allocatemem0(GEN z)
{
  ulong newsize;
  if (!z) newsize = 0;
  else {
    if (typ(z) != t_INT) pari_err_TYPE("allocatemem",z);
    newsize = itou(z);
    if (signe(z) < 0) pari_err_DOMAIN("allocatemem","size","<",gen_0,z);
  }
  allocatemem(newsize);
}

#include "gp_init.h"
#include "gp_default.h"

/* Backward Compatibility */

static GEN
gtype(GEN x)
{
  return stoi(typ(x));
}

static GEN
gsettype(GEN x,long t)
{
  x=gcopy(x); settyp(x,t); return x;
}

static long
setserieslength(long n)
{
  long m=precdl;
  if(n>0) precdl=n;
  return m;
}

static long
setprecr(long n)
{
  long m = GP_DATA->fmt->sigd;
  if (n > 0) { GP_DATA->fmt->sigd = n; precreal = ndec2prec(n); }
  return m;
}

entree functions_oldgp[] = {
{"allocatemem",11,(void *)allocatemem0,2,"vLp","allocatemem(s)=allocates a new stack of s bytes, or doubles the stack if size is 0"},
{"box",35,(void *)rectbox,10,"vLGG","box(w,x2,y2)=if the cursor is at position (x1,y1), draw a box with diagonal (x1,y1) and (x2,y2) in rectwindow w (cursor does not move)"},
{"color",2,(void *)rectcolor,2,"vLL","color(w,c)=set default color to c in rectwindow. Possible values for c are 1=sienna, 2=cornsilk, 3=red, 4=black, 5=grey, 6=blue, 7=gainsborough"},
{"cursor",11,(void*)rectcursor,10,"vLp","cursor(w)=current position of cursor in rectwindow w"},
{"default",0,(void*)default0,11,"D\"\",r,D\"\",s,D0,L,","default({opt},{v},{flag}): set the default opt to v. If v is omitted, print the current default for opt. If no argument is given, print a list of all defaults as well as their values. If flag is non-zero, return the result instead of printing it on screen. See manual for details"},
{"draw",1,(void*)rectdraw,10,"vGp","draw(list)=draw vector of rectwindows list at indicated x,y positions; list is a vector w1,x1,y1,w2,x2,y2,etc..."},
{"initrect",34,(void*)initrect,10,"vLLL","initrect(w,x,y)=initialize rectwindow w to size x,y"},
{"kill",85,(void*)kill0,11,"vr","kill(x)=kills the present value of the variable or function x. Returns new value or 0"},
{"killrect",11,(void *)killrect,10,"vL","killrect(w)=erase the rectwindow w"},
{"line",35,(void *)rectline,10,"vLGG","line(w,x2,y2)=if cursor is at position (x1,y1), draw a line from (x1,y1) to (x2,y2) (and move the cursor) in the rectwindow w"},
{"lines",35,(void *)rectlines,10,"vLGG","lines(w,listx,listy)=draws an open polygon in rectwindow w where listx and listy contain the x (resp. y) coordinates of the vertices"},
{"move",35,(void*)rectmove,10,"vLGG","move(w,x,y)=move cursor to position x,y in rectwindow w"},
{"plot",99,(void *)plot,10,"vV=GGEDGDGp","plot(X=a,b,expr)=crude plot of expression expr, X goes from a to b"},
{"ploth",37,(void *)ploth,10,"V=GGEp","ploth(X=a,b,expr)=plot of expression expr, X goes from a to b in high resolution"},
{"ploth2",37,(void *)ploth2,10,"V=GGEp","ploth2(X=a,b,[expr1,expr2])=plot of points [expr1,expr2], X goes from a to b in high resolution"},
{"plothmult",37,(void *)plothmult,10,"V=GGEp","plothmult(X=a,b,[expr1,...])=plot of expressions expr1,..., X goes from a to b in high resolution"},
{"plothraw",2,(void *)plothraw,10,"GGp","plothraw(listx,listy)=plot in high resolution points whose x (resp. y) coordinates are in listx (resp. listy)"},
{"point",35,(void *)rectpoint,10,"vLGG","point(w,x,y)=draw a point (and move cursor) at position x,y in rectwindow w"},
{"points",35,(void *)rectpoints,10,"vLGG","points(w,listx,listy)=draws in rectwindow w the points whose x (resp y) coordinates are in listx (resp listy)"},
{"postdraw",1,(void *)postdraw,10,"vG","postdraw(list)=same as plotdraw, except that the output is a PostScript program in file \"pari.ps\""},
{"postploth",37,(void *)postploth,10,"V=GGEpD0,L,D0,L,","postploth(X=a,b,expr)=same as ploth, except that the output is a PostScript program in the file \"pari.ps\""},
{"postploth2",37,(void *)postploth2,10,"V=GGEpD0,L,","postploth2(X=a,b,[expr1,expr2])=same as ploth2, except that the output is a PostScript program in the file \"pari.ps\""},
{"postplothraw",2,(void *)postplothraw,10,"GGD0,L,","postplothraw(listx,listy)=same as plothraw, except that the output is a PostScript program in the file \"pari.ps\""},
{"pprint",0,(void*)print,11,"vs*","pprint(a)=outputs a in beautified format ending with newline"},
{"pprint1",0,(void*)print1,11,"vs*","pprint1(a)=outputs a in beautified format without ending with newline"},
{"print",0,(void*)print,11,"vs*","print(a)=outputs a in raw format ending with newline"},
{"print1",0,(void*)print1,11,"vs*","print1(a)=outputs a in raw format without ending with newline"},
{"rbox",35,(void *)rectrbox,10,"vLGG","rbox(w,dx,dy)=if the cursor is at (x1,y1), draw a box with diagonal (x1,y1)-(x1+dx,y1+dy) in rectwindow w (cursor does not move)"},
{"read",0,(void *)input0,11,"","read()=read an expression from the input file or standard input"},
{"rline",35,(void *)rectrline,10,"vLGG","rline(w,dx,dy)=if the cursor is at (x1,y1), draw a line from (x1,y1) to (x1+dx,y1+dy) (and move the cursor) in the rectwindow w"},
{"rlines",35,(void *)rectlines,10,"vLGG","rlines(w,dx,dy)=draw in rectwindow w the points given by vector of first coordinates xsand vector of second coordinates, connect them by lines"},
{"rmove",35,(void *)rectrmove,10,"vLGG","rmove(w,dx,dy)=move cursor to position (dx,dy) relative to the present position in the rectwindow w"},
{"rpoint",35,(void *)rectrpoint,10,"vLGG","rpoint(w,dx,dy)=draw a point (and move cursor) at position dx,dy relative to present position of the cursor in rectwindow w"},
{"rpoints",35,(void *)rectpoints,10,"vLGG","rpoints(w,xs,ys)=draw in rectwindow w the points given by vector of first coordinates xs and vector of second coordinates ys"},
{"scale",59,(void *)rectscale,10,"vLGGGG","scale(w,x1,x2,y1,y2)=scale the coordinates in rectwindow w so that x goes from x1 to x2 and y from y1 to y2 (y2<y1 is allowed)"},
{"setprecision",15,(void *)setprecr,2,"lL","setprecision(n)=set the current precision to n decimal digits if n>0, or return the current precision if n<=0"},
{"setserieslength",15,(void *)setserieslength,2,"lL","setserieslength(n)=set the default length of power series to n if n>0, or return the current default length if n<=0"},
{"settype",21,(void *)gsettype,2,"GL","settype(x,t)=make a copy of x with type t (to use with extreme care)"},
{"string",57,(void*)rectstring,10,"vLs","string(w,x)=draw in rectwindow w the string corresponding to x, where x is either a string, or a number in R, written in format 9.3"},
{"system",70,(void*) system0,11,"vs","system(a): a being a string, execute the system command a (not valid on every machine)"},
{"texprint",0,(void*)printtex,11,"vs*","texprint(a)=outputs a in TeX format"},
{"type",1,(void *)gtype,2,"Gp","type(x)=internal type number of the GEN x"},

{NULL,0,NULL,0,NULL,NULL} /* sentinel */
,};
